torch/distributed/_tensor/experimental/attention.py

```diff
@@ -73,39 +83,70 @@ def _maybe_wait(tensor: torch.Tensor) -> torch.Tensor:
     return tensor
 
 
+def _partial_update(
+    original: torch.Tensor,
+    new: torch.Tensor,
+    dim: int,
+    n_chunks: int,
+    idx: int,
+    add: bool,
+) -> torch.Tensor:
+    chunks = list(original.chunk(n_chunks, dim=dim))
+    if add:
+        chunks[idx] += new
+    else:
+        chunks[idx] = new
+    return torch.cat(chunks, dim=dim)
+
+
 class _SDPAMerger:
     """A class to help to merge the local SDPA result."""
 
-    def __init__(self, convert_to_f32: bool):
+    def __init__(self, convert_to_f32: bool, seq_dim: int):
+        self._seq_dim = seq_dim
         self._out: Optional[torch.Tensor] = None
         self._lse: Optional[torch.Tensor] = None
         self._convert_to_f32 = convert_to_f32
         self._out_dtype = torch.float32
         self._lse_dtype = torch.float32
 
-    def _merge_one(self, block_out: torch.Tensor, block_lse: torch.Tensor) -> None:
+    def _merge_one(
+        self, block_out: torch.Tensor, block_lse: torch.Tensor, partial: bool
+    ) -> None:
         block_lse = block_lse.unsqueeze(dim=-1)
         if self._lse is None:
             self._lse = block_lse
             self._out = block_out
         else:
+            assert self._lse is not None
+            assert self._out is not None
+            lse = self._lse.chunk(2, dim=self._seq_dim)[1] if partial else self._lse
+            out = self._out.chunk(2, dim=self._seq_dim)[1] if partial else self._out
+
             # The algorithm from
             # github.com/zhuzilin/ring-flash-attention/pull/34#issuecomment-2076126795
             # gives a relatively stable result.
-            self._out = self._out - F.sigmoid(block_lse - self._lse) * (
-                self._out - block_out
+            new_lse = lse + torch.log(1 + torch.exp(block_lse - lse))
+            out = (
+                torch.exp(lse - new_lse) * out
+                + torch.exp(block_lse - new_lse) * block_out
             )
-            self._lse = self._lse - F.logsigmoid(self._lse - block_lse)
-
-    def step(self, out: torch.Tensor, lse: torch.Tensor) -> None:
+            if partial:
+                self._lse = _partial_update(self._lse, new_lse, 2, 2, 1, add=False)
+                self._out = _partial_update(self._out, out, 2, 2, 1, add=False)
+            else:
+                self._lse = new_lse
+                self._out = out
+
+    def step(self, out: torch.Tensor, lse: torch.Tensor, partial: bool) -> None:
```

Comment:
not for this PR,
i'm curious why we need to move lse to f32.  Should it be enough to move out to f32 (the first time we create out).  then we keep doing out += lse iiuc?  would it be ok to have f32 += bf16?

Reply:
oh nevermind. we don't accumulate lse into out. we accumulate partials into both of them.